File: src/java/net/dbyrne/natenberg/Natenberg.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package net.dbyrne.natenberg;

import java.io.IOException;

import net.sf.json.JSONException;

import com.ericsson.otp.erlang.OtpErlangDecodeException;
import com.ericsson.otp.erlang.OtpErlangExit;
import com.ericsson.otp.erlang.OtpMbox;
import com.ericsson.otp.erlang.OtpNode;

public class Natenberg {

	private OtpNode node = null;
	private OtpMbox mailbox = null;

	private final State start = new StateTransitionLogger(new StartState());
	private final State receiving = new StateTransitionLogger(new ReceivingState());
	private final State success = new StateTransitionLogger(new EndState(0));
	private final State fail = new StateTransitionLogger(new EndState(1));

	private final View view = new View();

	private interface State{ // FSM
		void transition();
	}

	public static void main(String[] args){
		new Natenberg().go();
	}

	public void go(){
		start.transition();
	}

	private class StartState implements State{
		public void transition() {
			try {
				node = new OtpNode("view@127.0.0.1");
				mailbox = node.createMbox("graph");
				receiving.transition();
			} catch (IOException e) {
				fail.transition();
			}
		}
	}

	private class ReceivingState implements State{
		private Long sequence = 1l;
		public void transition() {
			while(true){
				try {
					view.show(mailbox.receive(), sequence++);
					System.out.println("Message Processed");
				} catch (OtpErlangExit e) {
					success.transition();
				} catch (OtpErlangDecodeException e) {
					fail.transition();
				} catch (ClassCastException cce){
					cce.printStackTrace();
					view.show(cce.getClass().getName(), cce.getMessage(), sequence);
				} catch(JSONException je){
					je.printStackTrace();
					view.show(je.getClass().getName(), je.getMessage(), sequence);
				}
			}
		}
	}

	private class EndState implements State{
		private int returnValue;
		EndState(int returnValue){
			this.returnValue = returnValue;
		}
		public void transition() {
			if(mailbox != null)
				mailbox.close();
			if(node != null)
				node.close();
			System.exit(returnValue);
		}
	}

	private class StateTransitionLogger implements State{
		private final State wrapped;
		public StateTransitionLogger(State wrapped) {
			this.wrapped = wrapped;
		}
		public void transition() {
			String name = wrapped.getClass().getName();
			System.out.println("Entered " + name);
			wrapped.transition();
			System.out.println("Exit " + name);
		}
	}

}

File: src/erl/natenberg.erl
% Licensed under the Apache License, Version 2.0 (the "License"); you may not
% use this file except in compliance with the License.  You may obtain a copy of
% the License at
%
%   http://www.apache.org/licenses/LICENSE-2.0
%
% Unless required by applicable law or agreed to in writing, software
% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
% License for the specific language governing permissions and limitations under
% the License.

-module(natenberg).
-behaviour(gen_server).
% callbacks
-export([start/0, stop/0, start_link/0, init/1, handle_call/3, 
		 handle_cast/2, handle_info/2, terminate/2, code_change/3]).
-include_lib("struct.hrl").

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

start() ->
  application:start(natenberg).

stop() ->
  application:stop(natenberg).

start_link() ->
  gen_server:start_link({local, ?MODULE}, ?MODULE, self(), []).

init(_From) ->
  erlang:display(atom_to_list(?MODULE) ++ " starting ..."),
  process_flag(trap_exit,true),
  {ok, 0}.

handle_call(Function, _From, State) ->
  Json = erlang:apply(rest, Function, []),
  {reply, Json, State + 1}.

handle_cast(_Msg, State) ->
  {noreply, State}.
 
handle_info(_Info, State) ->
  {noreply, State}.

terminate(_Reason, _State) ->
  erlang:display(atom_to_list(?MODULE) ++ " terminating ..."),
  ok.

code_change(_Old, State, _Extra) ->
  {ok, State}.
